### PR TITLE
[go_router]Fixes GoRouterState.location and  GoRouterState.param to return correct value

### PR DESCRIPTION
**Pull request by chunhtai**
fixes https://github.com/flutter/flutter/issues/112444
fixes https://github.com/flutter/flutter/issues/111040

also cleans up RouteMatch and RouteMatchlist. This pr move all the properties in RouteMatch class that are the same for all of RouteMatch to RouteMatchlist to avoid duplicating data too many time.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [ ] I signed the [CLA].
- [ ] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by cgestes** [comment]
Does it also fixes wrong GoRouterState (accessed using GoRouteState.of(context)) when using router.push ?

**Comment by chunhtai** [comment]
> Does it also fixes wrong GoRouterState (accessed using GoRouteState.of(context)) when using router.push ?

I am not aware of the issue, is there an existing github issue? if not, can you file a new issue for this?

**Comment by cgestes** [comment]
here we go: https://github.com/flutter/flutter/issues/114901

File: packages/go_router/lib/src/delegate.dart [hunk 1]
```diff
@@ -185,3 +179,20 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
     return SynchronousFuture<void>(null);
   }
 }
+
+/// The route match that represent route pushed through [GoRouter.push].
+// TODO(chunhtai): Removes this once imperative API no longer insert route match.
+class ImperativeRouteMatch extends RouteMatch {
```
Comment (chunhtai) [inline packages/go_router/lib/src/delegate.dart:185]: This is a workaround that should be removed once the imperative API refactor is done

File: packages/go_router/lib/src/matching.dart [hunk 2]
```diff
@@ -188,59 +200,48 @@ List<RouteMatch> _getLocRouteRecursively({
...
-  if (result.isEmpty) {
-    return <RouteMatch>[];
+  if (result != null) {
+    pathParameters.addAll(subPathParameters);
   }
-
-  // If there are multiple routes that match the location, returning the first one.
-  // To make predefined routes to take precedence over dynamic routes eg. '/:id'
-  // consider adding the dynamic route at the end of the routes
-  return result.first;
```
Comment (chunhtai) [inline packages/go_router/lib/src/matching.dart:223]: This always returns first, might as well use a single object to hold the result. This is probably some left over code from previous refactoring.

File: packages/go_router/lib/src/delegate.dart [hunk 3]
```diff
@@ -95,26 +94,21 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   }
 
   /// Pushes the given location onto the page stack
-  void push(RouteMatch match) {
-    if (match.route is ShellRoute) {
-      throw GoError('ShellRoutes cannot be pushed');
-    }
+  void push(RouteMatchList matches) {
```
Comment (johnpryan) [inline packages/go_router/lib/src/delegate.dart:97]: `replace()` calls this method with a `RouteMatch`, not a `RouteMatchList` after https://github.com/flutter/packages/pull/2747. Should `replace()` also take a `RouteMatchList`?

**Comment by johnpryan** [comment]
@chunhtai I fixed the merge conflicts with master, but a few of the `replace()` tests are failing. Could you take a look at those?

File: packages/go_router/lib/src/builder.dart [hunk 4]
```diff
@@ -77,7 +78,7 @@ class RouteBuilder {
             return _buildErrorNavigator(
                 context,
                 e,
-                Uri.parse(matchList.location.toString()),
+                Uri.parse(matchList.uri.toString()),
```
Comment (ahmednfwela) [inline packages/go_router/lib/src/builder.dart:81]: can't this entire line be just `matchList.uri` ?

Comment (chunhtai) [reply packages/go_router/lib/src/builder.dart:81]: nice catch

**Review by johnpryan — APPROVED** [review]

**Comment by ahmednfwela** [comment]
@chunhtai are there any plans to expose `RouteMatchList` into the public API now that it's cleaned up ?

**Comment by chunhtai** [comment]
@ahmednfwela yes, that will be my next step. I will open up most of the internal classes.

**Pull request merged**